### PR TITLE
[MI-3719] Send the users an ephemeral message if they try to connect their account via MM desktop app

### LINKED ISSUE #398: Connecting to GitLab is failing from Mattermost desktop
Due to a change on GitLab's side (possibly related to Cloudflare - connecting to GitLab with the Mattermost desktop client is no longer working.

When you click the connection link, the modal opens but the spinner for `Checking if the site connection is secure` never stops and then the connection attempt eventrually times out.   
image

This is **not** a regression from the previous `1.6.0` release. The same behavior will occur there as well.

**Note**: Using Chrome or other browsers is not affected.

Steps:
- Install and enable the plugin
- Configure an oAuth app on GitLab side (or use Chimera)
- Run the `/gitlab connect` command and click the link

### PR DESCRIPTION
**Pull request by raghavaggarwal2308**
#### Summary
To avoid the indefinite loading state while connecting to Gitlab using the desktop app, we updated the logic to send the user a message and ask them to connect their account using the browser.

#### Screenshots:
Desktop app:
Screenshot from 2023-11-15 17-04-09

Browser:
image

#### What to test:
Users are getting an ephemeral message if they try to connect their account via MM desktop app

###### Steps to reproduce:
1. Open the Mattermost desktop app.
2. Make sure the Gitlab plugin is enabled.
3. Run the "/gitlab connect" command.

###### Environment:
MM version: v7.8.2
Node version: 14.18.0
Go version: 1.19.0

#### Ticket Link
Fixes #398

### CODE REVIEW

File: webapp/src/index.js [hunk 1]
```diff
@@ -38,6 +39,9 @@ class PluginClass {
         registry.registerBottomTeamSidebarComponent(TeamSidebar);
         registry.registerPopoverUserAttributesComponent(UserAttribute);
 
+        const hooks = new Hooks(store);
+        registry.registerSlashCommandWillBePostedHook(hooks.slashCommandWillBePostedHook);
```
Comment (mickmister) [inline webapp/src/index.js:43]: There are other places to handle for the connect case. Here's one:

https://github.com/mattermost/mattermost-plugin-gitlab/blob/7190bfda896caae089c36c0e7a65ca907ef1403c/webapp/src/components/sidebar_buttons/sidebar_buttons.jsx#L73

There may just be two, but I would look through the webapp directory to verify if there are more

Comment (raghavaggarwal2308) [reply webapp/src/index.js:43]: @mickmister We can connect our account using the RHS as well. We handled both cases now.

File: webapp/src/actions/index.js [hunk 2]
```diff
@@ -301,3 +305,26 @@ export function getChannelSubscriptions(channelId) {
...
+            id: 'gitlabPlugin' + Date.now(),
+            user_id: getState().entities.users.currentUserId,
+            channel_id: getCurrentChannelId(getState()),
+            message,
+            type: 'system_ephemeral',
+            create_at: timestamp,
+            update_at: timestamp,
+            root_id: '',
+            parent_id: '',
+            props: {},
```
Comment (avas27JTG) [inline webapp/src/actions/index.js:322]: if these are empty can we remove them?

File: webapp/src/hooks/index.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import {sendEphemeralPost} from '../actions';
+
+type ContextArgs = {channel_id: string};
+
+const connectCommand = '/gitlab connect';
+
+export default class Hooks {
+    private store: any;
+
+    constructor(store: any) {
```
Comment (avas27JTG) [inline webapp/src/hooks/index.ts:15]: can we remove `any` and add the proper type?

File: webapp/src/hooks/index.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        if (!message) {
+            return Promise.resolve({message, args: contextArgs});
+        }
+
+        if (message.startsWith(connectCommand)) {
+            if (isDesktopApp()){
+                this.store.dispatch(sendEphemeralPost(connectUsingBrowserMessage));
+                return Promise.resolve({});
+            }
+        }
```
Comment (avas27JTG) [inline webapp/src/hooks/index.ts:34]:
```suggestion
        if (message.startsWith(connectCommand) && isDesktopApp()) {
this.store.dispatch(sendEphemeralPost(connectUsingBrowserMessage));
         return Promise.resolve({});
        }
```

File: webapp/src/hooks/index.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+const connectCommand = '/gitlab connect';
+
+export default class Hooks {
+    private store: any;
+
+    constructor(store: any) {
+        this.store = store;
+    }
+
+    slashCommandWillBePostedHook = (rawMessage: string, contextArgs: ContextArgs) => {
```
Comment (avas27JTG) [inline webapp/src/hooks/index.ts:19]: have we tested for all other commands?

File: webapp/src/utils/user_agent.js [hunk 6]
```diff
@@ -0,0 +1,5 @@
+const userAgent = window.navigator.userAgent;
+
+export function isDesktopApp() {
+    return userAgent.indexOf('Mattermost') !== -1 && userAgent.indexOf('Electron') !== -1;
+}
```
Comment (avas27JTG) [inline webapp/src/utils/user_agent.js:5]:
```suggestion
export const isDesktopApp = () =>  userAgent.indexOf('Mattermost') !== -1 && userAgent.indexOf('Electron') !== -1;
```

File: webapp/src/actions/index.js (hunk 2 above)
Comment (raghavaggarwal2308) [reply webapp/src/actions/index.js:244]: @avas27JTG Those are mandatory fields

File: webapp/src/hooks/index.ts (hunk 5 above)
Comment (raghavaggarwal2308) [reply webapp/src/hooks/index.ts:19]: Yes, the other commands are not affected

File: webapp/src/actions/index.js [hunk 7]
```diff
@@ -223,3 +227,26 @@ export function getChannelSubscriptions(channelId) {
...
     };
 }
+
+export function sendEphemeralPost(message) {
+    return (dispatch, getState) => {
+        const timestamp = Date.now();
+        const post = {
+            id: 'gitlabPlugin' + Date.now(),
+            user_id: getState().entities.users.currentUserId,
+            channel_id: getCurrentChannelId(getState()),
```
Comment (mickmister) [inline webapp/src/actions/index.js:237]:
```suggestion
        const state = getState();

        const post = {
            id: 'gitlabPlugin' + Date.now(),
            user_id: getCurrentUserId(state),
            channel_id: getCurrentChannelId(state),
```

File: webapp/src/constants/index.js [hunk 8]
```diff
@@ -10,3 +10,5 @@ export const RHSStates = {
     UNREADS: 'unreads',
     ASSIGNMENTS: 'assignments',
 };
+
+export const connectUsingBrowserMessage = 'Mattermost desktop client does not support authenticating between Gitlab and Mattermost directly. To connect your Gitlab account with Mattermost, please go to Mattermost via your web browser and type `/gitlab connect`.';
```
Comment (mickmister) [inline webapp/src/constants/index.js:14]:
```suggestion
export const connectUsingBrowserMessage = 'Mattermost desktop client does not support authenticating between Gitlab and Mattermost directly. To connect your Gitlab account with Mattermost, please log in to Mattermost via your web browser and type `/gitlab connect`.';
```

**Review by mickmister — APPROVED** [review]
LGTM, just a few suggestions

**Comment by raghavaggarwal2308** [comment]
@mickmister Fixed the review comments mentioned by you.

**Comment by hanzei** [comment]
@raghavaggarwal2308 Would you please resolve the conflicts?

**Comment by raghavaggarwal2308** [comment]
@hanzei Fixed conflicts

**Comment by hanzei** [comment]
@AayushChaudhary0001 ready for your review

**Review by avas27JTG — APPROVED** [review]

**Comment by AayushChaudhary0001** [comment]
Tested and approved, LGTM

**Comment by raghavaggarwal2308** [comment]
@mickmister It will take some time to explore and see if this is possible. I think we should merge this PR and we can work on #418 as soon as we get time. Please let me know your opinions on this.

**Pull request merged**